Library/Homebrew/cask/cask_loader.rb

```diff
@@ -299,7 +299,12 @@ def load(config:)
 
         json_cask = Homebrew::API.merge_variations(json_cask).deep_symbolize_keys.freeze
 
-        cask_options[:tap] = Tap.fetch(json_cask[:tap]) if json_cask[:tap].to_s.include?("/")
+        cask_options[:tap] = if json_cask.key?(:tap) # public JSON v2
+          tap_value = json_cask.fetch(:tap)
+          Tap.fetch(tap_value) if tap_value.to_s.include?("/")
+        else # internal JSON v3
+          CoreCaskTap.instance
+        end
```

Comment:
This is just me trying to avoid using `Homebrew::API.internal_json_v3?` since we'll want to remove that eventually.

Reply:
I wonder if it makes sense to add the tap string to the `json_cask` value so that the cask gets saved with more information on install. It might make sense to include renames too and there might be something else I'm missing. This a cask-specific problem since we always store installed formulae with their source files.

Reply:
I'm not sure either of them are really necessary for uninstalls though.

Reply:
It'd be good to figure out what the bare minimum is that we actually need stored here for e.g. uninstalls and just store that, probably as a new dedicated JSON file that bears no resemblance to public or private APIs and never store the cask file.

Reply:
I don't think it makes sense to work on and have to maintain another format. We should be good as long as tap is included so maybe we should just add it in the be sure about things. I'm going to think a bit more about how we want to handle loading legacy cask JSON that didn't have the tap defined.

My first thought is that if we're loading from a JSON blob or file we should just use the defined tap if it exists.When loading from the API it should probably just default to the core cask tap.

Reply:
> I don't think it makes sense to work on and have to maintain another format.

I don't think it makes sense to store a bunch of data here that's not actually used and reuse/tweak an existing format when we already have two different "formats" (cask file vs. JSON on disk).

Note that this has been a pretty long-running issue in Cask due to writing too much here and breaking backwards compatibility in the past causing Casks to not be able to be uninstalled.

What I'm arguing for is essentially:
- reusing the internal JSON format for this is not a good idea and going to cause more issues in future (as it has in the past)
- the existing format contains far too much unnecessary data which has and will continue to cause issues due to the too-large surface area here that's too easy for us to break with deprecation/disables/removals
- having 2-3 paths for what files are stored in the Caskroom is bad when Formulae has only one
- less of a big deal but: having how Formulae and Casks behave so differently in this way is undesirable

Genuinely I apologise that I keep stumbling on things that end up blowing up the scope here, it's just that some of this stuff has been _very_ painful over long periods of time in Homebrew and I'd rather we did it properly if we're going to introduce a new API for this.

Genuinely appreciate all your work here and elsewhere: thank you @apainintheneck ❤️ 

> My first thought is that if we're loading from a JSON blob or file we should just use the defined tap if it exists.When loading from the API it should probably just default to the core cask tap.

This seems reasonable.

I can't see when, with the existing code, how JSON would be used for non-core cask taps though, right? It'll always use the Cask file I'd think.

Reply:
So what would the one file type be in the case that we decided to only support one format for installed casks? We could download the original cask file whenever we install a cask and just store that. Then, we wouldn't have to support the installed JSON file at all. Or we could generate a JSON file for every cask install and only store the JSON representation.

Reply:
> So what would the one file type be in the case that we decided to only support one format for installed casks?

A JSON file.

> Or we could generate a JSON file for every cask install and only store the JSON representation.

A minimal JSON file that's a subset of internal JSON v3 containing only the information needed to e.g. uninstall a cask.